### ISSUE TITLE
Failing suspected API request to not-running server: /hub/user/<user>/metrics

### ISSUE THREAD

**Issue by masaok** [jupyterhub/the-littlest-jupyterhub#427]
```DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.2 LTS"
```
I'm getting tons of logs in `/var/log/syslog` that look like this:

```
Sep  5 23:15:55 jupyter1 python3[791]: [I 2019-09-05 23:15:55.823 JupyterHub log:174] 302 GET /user/cgalvin/metrics -> /hub/user/studentuser/metrics (@XXX.XXX.XXX.XXX) 2.40ms
Sep  5 23:15:55 jupyter1 python3[791]: [W 2019-09-05 23:15:55.854 JupyterHub base:1175] Failing suspected API request to not-running server: /hub/user/studentuser/metrics
```

However, when I log in via a browser and check `/user/studentuser/metrics`, it returns correctly: 
```
{
rss: 122232832,
limits: { }
}
```
With a `200` in the logs.

I've tried this as an admin user and normal user, and everything looks normal in the logs, so I'm not sure what is causing this error when other users are using the site.

**Comment by lahwaacz** [comment]
I'm getting a _lot_ of similar logs on JupyterHub 4.0.2 for the `/jupyter/hub/user/<username>/api/kernels/<uuid>` endpoint:
```
[I 2023-09-27 23:13:22.744 JupyterHub log:191] 302 GET /jupyter/user/<username>/api/kernels/<uuid>?1695849202650 -> /jupyter/hub/user/<username>/api/kernels/<uuid>?1695849202650 (@<ip>) 0.90ms
[W 2023-09-27 23:13:22.754 JupyterHub base:1444] Failing suspected API request to not-running server: /jupyter/hub/user/<username>/api/kernels/<uuid>
[W 2023-09-27 23:13:22.755 JupyterHub log:191] 424 GET /jupyter/hub/user/<username>/api/kernels/<uuid>?1695849202650 (<username>@<ip>) 3.76ms
```

I'm fairly certain the users don't run any scraper or something, so where does this come from?

**Comment by minrk** [comment]
@lahwaacz yes, most likely is that a user has a JupyterLab tab open after their server has been shutdown. JupyterLab is known to be overzealous at polling the server, especially after it's become unavailable.

The /metrics endpoint is used by [jupyter-resource-usage](https://github.com/jupyter-server/jupyter-resource-usage) (it's been deprecated quite some time ago in favor of `/metrics/v1/...`), so some packages might be out of date).

This particular poll is [created here](https://github.com/jupyter-server/jupyter-resource-usage/blob/cbf962f881058a30e2ce6951018da1cc937b8929/packages/labextension/src/model.ts#L40C21-L42), with a _max_ poll interval unspecified, so it uses the [default of 30 seconds](https://github.com/jupyterlab/lumino/blob/171c4f62fb61a1a72215a51c196f2f5a7d8b1186/packages/polling/src/poll.ts#L458). That means that while a server is stopped, you will get these errors every 30 seconds from every browser left open pointing to a not-running server. It should only poll when the window is focused, though, so it's a bit weird that you're seeing so many, since that should suggest there are _focused_ JupyterLab windows open and in the foreground on computers with no running server behind them. If the focus control isn't happening, it may be because things are out of date.

jupyter-resource-metrics could set a longer max poll interval so you get these less often, or it could add some logic to stop polling altogether after too many failures, waiting for a manual restart.

**Comment by lahwaacz** [comment]
@minrk Can you point to some JupyterLab issues about this behavior?

**Comment by minrk** [comment]
https://github.com/jupyterlab/jupyterlab/issues/3929 is the main one

Can you share your user and hub environments? I don't think `/hub/user/metrics` will be hit at all if packages are up to date, so maybe updating will fix the polling problem as well.

**Comment by lahwaacz** [comment]
@minrk I'm not getting hits on `/hub/user/<user>/metrics`, only `/jupyter/hub/user/<username>/api/kernels/<uuid>`.

I run JupyterHub on Arch Linux where the package names and versions are the following (most likely different from `pip` or other package managers):
```
jupyter-bash_kernel 0.9.0-2
jupyter-gnuplot_kernel 0.4.1-6
jupyter-lsp 2.2.0-2
jupyter-metakernel 0.30.1-1
jupyter-nbclient 0.8.0-1
jupyter-nbconvert 7.8.0-1
jupyter-nbformat 5.9.2-1
jupyter-nbgitpuller 1.2.0-2
jupyter-nbgrader-git v0.9.1.r1.g2ef44515-1
jupyter-notebook 7.0.4-1
jupyter-notebook-shim 0.2.3-2
jupyter-server 2.7.3-1
jupyterhub 4.0.2-1
jupyterhub-idle-culler 1.2.1-2
jupyterhub-systemdspawner 1.0.1-1
jupyterlab 4.0.6-1
jupyterlab-pygments 0.2.2-1
jupyterlab-pytutor 0.2.0-2
jupyterlab-rise 0.41.0-1
python-jupyter-client 8.3.1-1
python-jupyter-core 5.3.2-1
python-jupyter-events 0.7.0-1
python-jupyter-server-terminals 0.4.4-4
python-jupyter_telemetry 0.1.0-2
python-jupyterlab-server 2.25.0-1
```

**Comment by minrk** [comment]
Sorry, I mixed your report and the OP which reported /metrics.

The kernels list is definitely jupyterlab, though I can't say for certain that it's not in one of the extensions. It's definitely _supposed_ to stop polling when connection is lost, so I think this is a bug in jupyterlab that polling continues when the connection is lost. I can't say for sure where, but it's worth reporting to jupyterlab.

references:

- standby (which pauses polling) is meant to de set [when disconnected or hidden](https://github.com/jupyterlab/jupyterlab/blob/6fa51006d984cbf5489329eb3c552f5ed67b4ae2/packages/application/src/lab.ts#L29)
- special jupyterhub handling of disconnect is [here](https://github.com/jupyterlab/jupyterlab/blob/6fa51006d984cbf5489329eb3c552f5ed67b4ae2/packages/hub-extension/src/index.ts#L152C19-L190)

**Comment by lahwaacz** [comment]
@minrk Thanks, I reported it in https://github.com/jupyterlab/jupyterlab/issues/15211. Sorry for mixing possibly unrelated issue here.

**Comment by minrk** [comment]
No worries, I think it's the same general issue, just the exact manifestation may change over time.
